lib/ansible/modules/pip.py

```diff
@@ -611,6 +613,64 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err, cmd
 
 
+def _resolve_package_names(
+        module: AnsibleModule,
+        package_list: list[Package],
+        pip: list[str]
+) -> list[Package]:
+    """Return a list of Packages with names extracted from metadata using pip"""
+    pkgs_to_resolve = [pkg for pkg in package_list if not pkg.has_requirement]
+
+    if not pkgs_to_resolve:
+        return package_list
+
+    # Unsetting 'FORCE_COLOR' with 'NO_COLOR' and 'TTY_COMPATIBLE' env vars, helps to
+    # make pip output clean JSON without injected ANSI sequences that might corrupt it.
+    os.environ.pop('FORCE_COLOR', None)
+
+    rc, json_out, err = module.run_command(
+        [*pip, 'install', '--dry-run', '--ignore-installed', '--quiet', '--report=-', *(str(pkg) for pkg in pkgs_to_resolve)],
+        environ_update={
+            'NO_COLOR': '1',
+            'TTY_COMPATIBLE': '0',
+        },
+    )
+
+    if rc == 0:
+        # If it succeeds, handle the report
+        report = json.loads(json_out)
+        package_objects = (Package(install_report['metadata']['name'], version_string=install_report['metadata']['version'])
+                           for install_report in report['install'])
+    else:
+        # Else, if that fails due to --dry-run not being present in pip
+        # versions older than 22.2 (ex: Ubuntu 22.04), fall back to using
+        # the potentially faulty pip download method and warn the user.
+
+        pip_downloads_dir = Path(module.tmpdir) / 'pipdownloads'
+
+        rc, out, err = module.run_command([*pip, 'download', f'--dest={pip_downloads_dir}', '--no-deps',
+                                           *(pkg.download_information for pkg in pkgs_to_resolve)])
+
+        if rc != 0:
+            # If it fails (usually due to no pkg at url), just dump the error
+            module.fail_json(rc=rc, msg=out, err=err)
+
+        # Only from this point onward can we be sure that we're using the fallback, warn now.
+        module.warn("Using check mode with URL-sourced packages is potentially error-prone on pip versions older than 22.2")
+
+        out_lines = out.splitlines()
+
+        saved_packages = [Path(line.rsplit(" ", 1)[1]) for line in out_lines if 'Saved' in line]
+
+        if saved_packages:
+            package_objects = (Package.from_dist_path(path) for path in saved_packages)
+
+        package_objects = [Package(line.rsplit(" ", 1)[1]) for line in out_lines if 'Successfully downloaded' in line]
```

Comment:
You're first declaring the var a few lines above. And that definition doesn't have explicit MyPy typing info. So MyPy infers it from that first assignment, which is a generator expression. And any variable use after that is expected to have that type. Here, you attempted assigning a regular list which is an incompatible type (as far as type checkers are concerned: this doesn't error out in actual runtime).

This can be fixed by using a generator expression here as well (I don't see a reason for a list anyway):
```suggestion
        package_objects = (
            Package(line.rsplit(" ", 1)[1]) for line in out_lines
            if 'Successfully downloaded' in line
        )
```

Reply:
No longer in the code. Good to know on the MyPy stuff though.